server/app/mutations/grid_certificates/common.rb

```diff
@@ -64,5 +69,84 @@ def validate_dns_record(domain, expected_record)
     rescue
       false
     end
+
+    # @param domain_authorization [GridDomainAuthorization]
+    def validate_domain_authorization(authz)
+
+      case authz.authorization_type
+      when 'dns-01'
+        # Check that the expected DNS record is already in place
+        unless check_dns_record(authz.domain, authz.challenge_opts['record_content'])
+          add_error(:dns_record, :invalid, "Expected DNS record not present for domain #{authz.domain}") # XXX: validations error type
+        end
+      end
+    end
+
+    # @param domains [Array<String>]
+    def validate_authorizations_for_domains(domains)
+      domains.each do |domain|
+        unless domain_authorization = get_authz_for_domain(self.grid, domain)
+          add_error(:authorization, :not_found, "Domain authorization not found for domain #{domain}") # XXX: validations error type
+          return # No point to continue validations
+        end
+
+        validate_domain_authorization(domain_authorization)
+      end
+    end
+
+    # @param grid [Grid]
+    # @param le_client [Acme::Client] for grid
+    # @param domain [String]
+    # @raise [Timeout::Error]
+    def verify_domain(grid, le_client, domain)
+      domain_authorization = get_authz_for_domain(grid, domain)
+      challenge = le_client.challenge_from_hash(domain_authorization.challenge)
+      if domain_authorization.state == :created
+        info "requesting verification for domain #{domain}"
+        success = challenge.request_verification
+        if success
+          domain_authorization.state = :requested
+          domain_authorization.save
+        else
+          add_error(:request_verification, :failed, "Requesting verification failed")
+        end
+      end
+
+      wait_until!("domain verification for #{domain} is valid", interval: 1, timeout: 30, threshold: 10) {
+        challenge.verify_status != 'pending'
+      }
+
+      domain_authorization.expires = nil # XXX: old version of acme-client does not update authorization expiration
```

Comment:
The old 0.3.6 acme-client with `Acme::Client::Resources::Challenges::*.verify_status` does not allow us to see the updated `Acme::Client::Resources::Authorization.expires`, which changes when the ACME domain authorization goes from `pending` => `valid`.

The newer 0.6 version replaces the `challenge.verify_status` with [`authorization.verify_status`](https://github.com/unixcharles/acme-client/blob/bc5688bdc405053a864b7b50a22dee9b55bb6e05/lib/acme/client/resources/challenges/base.rb#L13) which does [update](https://github.com/unixcharles/acme-client/blob/bc5688bdc405053a864b7b50a22dee9b55bb6e05/lib/acme/client/resources/authorization.rb#L17) the `@expired` attribute.

That would be necessary for us to know when the validated domain authz expires. that would be relevant if we later re-request/renewing certificates for the same domains: the domains don't need to be re-authorized if the previous domain authz is still valid.

Reply:
Could we update `acme-client`?

Reply:
There's other breaking changes in the newer acme-client: https://github.com/kontena/kontena/pull/2565#issuecomment-321798538

The way that the `GridDomainAuthorization.challenge` has is used no long works... the model needs to be changed to store the `authorization.uri` for `client.fetch_authorization(...)` instead. Not something I want to do in a 1.4.x patch release...

That change also isn't strictly relevant for the fix here... only if we want to track `expires` for validated domain authorizations.